scripts/format_notebooks.py

```diff
@@ -289,6 +291,50 @@ def preprocess(self, nb, resources):
         return nb, resources
 
 
+class IdempotentIdPreprocessor(preprocessors.Preprocessor):
+    # https://github.com/jupyter/enhancement-proposals/blob/master/62-cell-id/cell-id.md introduces
+    # 'cell ids', which nbformat 5.1.0+ inserts. However, it inserts random ones. This class
+    # overwrites the random ones with fixed IDs.
+
+    KEY = "id-source-seen"
+
+    def preprocess(self, nb, resources):
+        # a scoped dictionary to track the source code seen in this notebook
+        resources[self.KEY] = {}
+        out_nb, out_resources = super().preprocess(nb, resources)
+        out_resources.pop(self.KEY)
+
+        # double check the IDs are actually assigned uniquely
+        ids = [c.id for c in out_nb.cells]
+        assert len(ids) == len(set(ids))
+
+        return out_nb, out_resources
+
+    def preprocess_cell(self, cell, resources, cell_index):
+        cell = copy.deepcopy(cell)
+        source = cell["source"]
+
+        # the core of the ID is a hash of the source code: this will be the same across every time a
+        # given version of the notebook is reformatted.
```

Comment:
This comment is a little confusing - you mean this will generate the same IDs every time it's run on the same input notebook, not that it's idempotent when the notebook is reformatted?

Reply:
This preprocessor is individually idempotent because it's not editing the source, and `format_notebooks.py --default` as a whole is idempotent because each cell's source is formatted before computing these ids, and so it's using the same source as it would when run a second time (where that second time wouldn't have to reformat individual cells).